### ISSUE TITLE
Database refresh not refreshing

### ISSUE THREAD

**Issue by spawn-github** [ethanopp/fitly#15]
Hi! Having trouble refreshing the database.  So far, I don't believe I've been able to import any data from any of the connected apps (strava, oura, or withings) I did pull the docker image tonight, so I'm on the latest code for this report

At first, when trying to refresh the database, I was getting a message to the effect that my athlete data was incomplete (that was a few days ago, so I don't have the precise errors). When saving each item of athlete data, all seemed well, except for 'recovery metric'. Each time I saved that setting, no matter the chosen value, I was seeing something close to this in the output:

`Error resetting hrv workout plan: unsupported operand type(s) for +: 'NoneType' and 'datetime.timedelta'`
or
`Error resetting hrv workout plan: 'NoneType' object has no attribute 'date'`

The chosen recovery metric value always updates as expected though. There was, and still is, a red outline around the athlete box, so maybe that's a clue?

At some point, the athlete data error went away when refreshing, and now attempting to refresh data launches the `Database Refresh in Progress` modal, but it just sits there spinning away, never going away on it's own. None of the dashboards are populating, so I'm presuming it's not actually pulling data. The only errors I see logged when refreshing are the above hrv workout plan errors, so not sure where to go from here.

Again, happy to provide any additional info.

Thanks!

**Comment by ethanopp** [comment]
Ah yeah sorry about that - I’ve been making a lot of updates which use that recovery metric over the last week so there may be some bugs. Will try to take a look at the code for this later today.

What I think might be happening though is that when you save the recovery metric, the system tries to query your oura data to generate workout recommendations, but if you’re doing that before loading any oura data in it’ll get stuck trying to query that data.

For now, try this:
1. Delete fitness.db file and restart fitly
2. Click “Debug” in the logs section and connect all 3rd party apps and ensure the top left says “Strava/Oura/Withings Connected”
3. Update all athlete settings EXCEPT recovery metric
4. Click refresh button

Let me know what log section shows

**Comment by spawn-github** [comment]
Following the steps, saving all except recovery metric, yields the following when refreshing (saved values sanitized):

`Updated name to Foo
Updated birthday to 1970-01-01 00:00:00
Updated sex to M
Updated weight_lbs to 200
Updated resting_hr to 100
Updated ride_ftp to 0
Updated run_ftp to 0
Manually refreshing database tables...
Please define all athlete settings prior to refreshing data`

Appreciate the assistance!

**Comment by ethanopp** [comment]
Latest commit should now deal with the modal hanging and `hrv workout plan` issue...

In regards to the athlete values, I think the code is interpreting you ftp values as 0 as None, so it thinks you never saved them... Could you try putting actual values for ride_ftp and run_ftp?

**Comment by spawn-github** [comment]
That makes sense, and I can confirm that before pulling the latest image,  I updated the ftp values to non-zero, and the data refresh started.

I've pulled the latest image, and the data refresh is mostly working (I have data in the dashboards now!), except for what I'm guessing are bigger strava activities:

`Activity id "4463824670": Assigning athlete id 1
Activity id "4463824670": Building df_samples
Activity id "4463824670": Building df_summary
Activity id "4463824670": Pulling ftp
Activity id "4463824670": Pulling resting hr
Activity id "4463824670": Pulling weight
Activity id "4463824670": Calculating power zones
Activity id "4463824670": Calculating heartrate zones
Activity id "4463824670": Calculating zones intensities
Activity id "4463824670": Calculating summary analytics
Activity id "4463824670": Writing mean max power to DB
[2021-01-09 19:17:00 +0000] [1] [CRITICAL] WORKER TIMEOUT (pid:16)
Error pulling strava data: 1`

I was seeing worker timeouts with some other activities as well, but re-running the refresh would eventually get through the processing. With this particular activity, after ~5 refreshes, I'm not able to get the activity processed.

The hanging modal is mostly resolved as well, but it appears an automated refresh occurs from time to time, and if a manual refresh was already in process, the modal ends up hanging. 

Thanks again for the quick fixes!
